packages/node/src/behavior/system/subscription/SubscriptionBehavior.ts

```diff
@@ -177,17 +177,23 @@ export class SubscriptionBehavior extends Behavior {
         this.state.subscriptions.push(peerSubscription);
     }
 
-    #subscriptionCancelled(subscription: Subscription) {
+    #subscriptionCancelled(subscription: Subscription): MaybePromise {
         if (subscription.isCanceledByPeer) {
             if (this.state.persistenceEnabled === false) return;
             const { id } = subscription;
             const subscriptionIndex = this.state.subscriptions.findIndex(({ subscriptionId }) => id === subscriptionId);
             if (subscriptionIndex !== -1) {
-                this.state.subscriptions.splice(subscriptionIndex, 1);
+                return this.#removeSubscriptionIndex(subscriptionIndex);
             }
         }
     }
 
+    async #removeSubscriptionIndex(index: number) {
+        await this.context.transaction.begin();
```

Comment:
If this is working great, you may need to add the resource too to actually grab the lock. Something like `this.transaction.addResource(this)`

Reply:
adjusted